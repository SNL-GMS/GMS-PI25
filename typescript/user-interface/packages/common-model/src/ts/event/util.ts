import type { Event, EventHypothesis, LocationSolution } from './types';

/**
 * Finds preferred event hypothesis for a given event and stage
 * If no preferred hypothesis exists for stage, will return most recent preferred hypothesis by stage
 *
 * @param event Event to search within
 * @param openIntervalName Currently open interval/stage name eg; "AL1"
 */
export const findPreferredEventHypothesisByStage = (
  event: Event,
  openIntervalName: string
): EventHypothesis => {
  if (!event) {
    return undefined;
  }

  const { preferredEventHypothesisByStage, eventHypotheses } = event;
  if (preferredEventHypothesisByStage?.length === 0) {
    return undefined;
  }

  const preferredHypoByStageId =
    preferredEventHypothesisByStage.find(hypo => hypo.stage.name === openIntervalName)?.preferred ??
    preferredEventHypothesisByStage[0].preferred;

  return eventHypotheses.find(
    hypothesis => hypothesis.id.hypothesisId === preferredHypoByStageId.id.hypothesisId
  );
};

/**
 * Determines if the provided {@link EventHypothesis} object is the
 * preferredEventHypothesis for the current open stage.
 *
 * @param event Event object to check
 * @param openIntervalName Current open stage
 * @param eventHypothesis EventHypothesis object to verify
 * @returns true if the eventHypothesis is preferred
 */
export const isPreferredEventHypothesisByStage = (
  event: Event,
  openIntervalName: string,
  eventHypothesis: EventHypothesis
): boolean => {
  if (!event || !eventHypothesis) {
    return false;
  }

  // Get the preferredEventHypo exists for the open stage (if it exists)
  const maybePreferred = event.preferredEventHypothesisByStage.find(
    hypo => hypo.stage.name === openIntervalName
  );

  // Check if these two have the same identifiers
  return (
    maybePreferred?.preferred.id.eventId === eventHypothesis.id.eventId &&
    maybePreferred?.preferred.id.hypothesisId === eventHypothesis.id.hypothesisId
  );
};

/**
 * Finds the last non rejected parent event hypothesis for a given event and hypothesis
 * if no valid hypothesis is found it returns the first hypothesis
 *
 * @param event
 * @param eventHypothesis
 */
export const findEventHypothesisParent = (
  event: Event,
  eventHypothesis: EventHypothesis
): EventHypothesis | undefined => {
  if (!eventHypothesis) {
    return undefined;
  }

  // loop backwards until we find a non-rejected hypothesis
  for (let i = eventHypothesis.parentEventHypotheses.length - 1; i >= 0; i -= 1) {
    const parentEventHypothesis = event.eventHypotheses.find(
      hypo => hypo.id.hypothesisId === eventHypothesis.parentEventHypotheses[i].id.hypothesisId
    );
    if (parentEventHypothesis) {
      return parentEventHypothesis;
    }
  }

  // nothing was found, return the first hypothesis
  return event.eventHypotheses[0];
};

/**
 * Finds the preferred location solution for a hypothesis falling back to the parents if the hypothesis is rejected
 *
 * @param eventHypothesisId hypothesis id to find the solution for
 * @param eventHypotheses list of hypotheses, if an event is opened
 * @returns a location solution
 */
export const findPreferredLocationSolution = (
  eventHypothesisId: string,
  eventHypotheses: EventHypothesis[]
): LocationSolution => {
  const eventHypothesis = eventHypotheses.find(
    hypothesis => hypothesis.id.hypothesisId === eventHypothesisId
  );
  if (!eventHypothesis) return undefined;
  if (eventHypothesis.preferredLocationSolution) {
    return eventHypothesis.locationSolutions.find(
      ls => ls.id === eventHypothesis.preferredLocationSolution.id
    );
  }

  if (eventHypothesis.parentEventHypotheses) {
    const parentHypothesisId =
      eventHypothesis.parentEventHypotheses[eventHypothesis.parentEventHypotheses.length - 1].id
        .hypothesisId;

    const parentEventHypothesis = eventHypotheses.find(
      hypothesis => hypothesis.id.hypothesisId === parentHypothesisId
    );

    return parentEventHypothesis.locationSolutions.find(
      ls => ls.id === parentEventHypothesis.preferredLocationSolution.id
    );
  }

  return undefined;
};
