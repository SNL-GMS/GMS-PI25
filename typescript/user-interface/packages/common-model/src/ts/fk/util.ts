import type {
  FkConfiguration,
  FkConfigurationDefaults,
  FkParams,
  FKPowerSpectrumDefinition,
  FrequencyBand,
  LeadLagPairAndString,
  WindowParameters
} from './types';
import { LeadLagPairs } from './types';

// TODO: Review to see what might make more sense as part of configuration
// Width/height of y and x axis respectively
export const SIZE_OF_FK_RENDERING_AXIS_PX = 35;

/**
 * Hard-coded lead/lag options
 */

export const LeadLagPairsValues: WindowParameters[] = [
  {
    leadSeconds: 1,
    lengthSeconds: 4,
    stepSize: undefined
  },
  {
    leadSeconds: 1,
    lengthSeconds: 6,
    stepSize: undefined
  },
  {
    leadSeconds: 1,
    lengthSeconds: 9,
    stepSize: undefined
  },
  {
    leadSeconds: 1,
    lengthSeconds: 11,
    stepSize: undefined
  }
];

export const LeadLagValuesAndDisplayString: LeadLagPairAndString[] = [
  {
    leadLagPairs: LeadLagPairs.LEAD_1_DURATION_4,
    windowParams: LeadLagPairsValues[0]
  },
  {
    leadLagPairs: LeadLagPairs.LEAD_1_DURATION_6,
    windowParams: LeadLagPairsValues[1]
  },
  {
    leadLagPairs: LeadLagPairs.LEAD_1_DURATION_9,
    windowParams: LeadLagPairsValues[2]
  },
  {
    leadLagPairs: LeadLagPairs.LEAD_1_DURATION_11,
    windowParams: LeadLagPairsValues[3]
  }
];

/** Default Frequency Bands */
export const FrequencyBands: FrequencyBand[] = [
  {
    minFrequencyHz: 0.5,
    maxFrequencyHz: 2
  },
  {
    minFrequencyHz: 1,
    maxFrequencyHz: 2.5
  },
  {
    minFrequencyHz: 1.5,
    maxFrequencyHz: 3
  },
  {
    minFrequencyHz: 2,
    maxFrequencyHz: 4
  },
  {
    minFrequencyHz: 3,
    maxFrequencyHz: 6
  }
];

/**
 * Approximate conversion between degrees and km
 */
export function degreeToKmApproximate(degree: number): number {
  const DEGREES_IN_CIRCLE = 360;
  const RAD_EARTH = 6371;
  const TWO_PI = Math.PI * 2;
  return degree / (DEGREES_IN_CIRCLE / (RAD_EARTH * TWO_PI));
}

/**
 * Approximate conversion between km and degrees
 */
export function kmToDegreesApproximate(km: number): number {
  const DEGREES_IN_CIRCLE = 360;
  const RAD_EARTH = 6371;
  const TWO_PI = Math.PI * 2;
  return km * (DEGREES_IN_CIRCLE / (RAD_EARTH * TWO_PI));
}

export const defaultFkPowerSpectrum: FKPowerSpectrumDefinition = {
  windowLead: 5,
  windowLength: 10,
  startTimeStep: 2,
  lowFrequency: 1.25,
  highFrequency: 3.25,
  useChannelVerticalOffsets: true,
  phaseType: 'P'
};

export const fkConfigurationDefaults: FkConfigurationDefaults = {
  defaultMaximumSlowness: 40,
  defaultNumberOfPoints: 81,
  defaultLead: 1,
  defaultLength: 4,
  defaultStepSize: 5,
  fkNeedsReviewRuleSet: {
    phasesNeedingReview: ['P', 'Pg', 'Pn']
  }
};

/**
 * Returns an empty FK Spectrum configuration. The values are NOT default values,
 * but instead values that will make it obvious within the UI that a correct
 * configuration was never added to the FK
 */
export const defaultFkConfiguration: FkConfiguration = {
  contributingChannelsConfiguration: [],
  maximumSlowness: fkConfigurationDefaults.defaultMaximumSlowness,
  mediumVelocity: 1,
  normalizeWaveforms: false,
  numberOfPoints: fkConfigurationDefaults.defaultNumberOfPoints,
  useChannelVerticalOffset: false,
  leadFkSpectrumSeconds: fkConfigurationDefaults.defaultLead
};

/**
 * Returns default FkParams for when no Fk is available
 *
 * @returns default populated FkParams
 */
export function getDefaultFkParams(): FkParams {
  return {
    frequencyPair: {
      minFrequencyHz: defaultFkPowerSpectrum.lowFrequency,
      maxFrequencyHz: defaultFkPowerSpectrum.highFrequency
    },
    windowParams: {
      leadSeconds: fkConfigurationDefaults.defaultLead,
      lengthSeconds: fkConfigurationDefaults.defaultLength,
      stepSize: fkConfigurationDefaults.defaultStepSize
    }
  };
}
