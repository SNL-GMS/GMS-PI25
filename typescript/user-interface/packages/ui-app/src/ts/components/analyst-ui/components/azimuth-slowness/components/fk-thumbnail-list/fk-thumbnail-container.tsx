/* eslint-disable react/destructuring-assignment */
import { NonIdealState } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import type { EventTypes, FkTypes } from '@gms/common-model';
import { SignalDetectionTypes } from '@gms/common-model';
import React from 'react';

import { FkThumbnail } from '../fk-thumbnail';
import * as fkUtil from '../fk-util';

/**
 * Fk Thumbnail Props.
 */
export interface FkThumbnailContainerProps {
  data: SignalDetectionTypes.SignalDetection;
  signalDetectionFeaturePredictions: EventTypes.FeaturePrediction[];
  sizePx: number;
  selected: boolean;
  isUnassociated: boolean;
  fkUnit: FkTypes.FkUnits;
  arrivalTimeMovieSpectrumIndex: number;
  selectedFk: FkTypes.FkPowerSpectra;
  showFkThumbnailMenu?(event: React.MouseEvent): void;
  onClick?(e: React.MouseEvent<HTMLCanvasElement>): void;
}

/**
 * A single fk thumbnail in the thumbnail-list
 */
export function FkThumbnailContainer(props: FkThumbnailContainerProps) {
  if (!props.data) {
    return <NonIdealState icon={IconNames.HEAT_GRID} title="All Fks Filtered Out" />;
  }
  const fmPhase = SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(
    SignalDetectionTypes.Util.getCurrentHypothesis(props.data.signalDetectionHypotheses)
      .featureMeasurements
  );

  const needsReview = fkUtil.fkNeedsReview(props.data, props.selectedFk);
  const label = `${props.data.station.name} ${fmPhase.value.toString()}`;
  const predictedPoint = fkUtil.getPredictedPoint(props.signalDetectionFeaturePredictions);

  return (
    <FkThumbnail
      fkData={props.selectedFk}
      label={label}
      dimFk={props.isUnassociated}
      highlightLabel={needsReview}
      fkUnit={props.fkUnit}
      sizePx={props.sizePx}
      // eslint-disable-next-line @typescript-eslint/unbound-method
      onClick={props.onClick}
      predictedPoint={predictedPoint}
      selected={props.selected}
      // eslint-disable-next-line @typescript-eslint/unbound-method
      showFkThumbnailMenu={props.showFkThumbnailMenu}
      arrivalTimeMovieSpectrumIndex={props.arrivalTimeMovieSpectrumIndex}
    />
  );
}
