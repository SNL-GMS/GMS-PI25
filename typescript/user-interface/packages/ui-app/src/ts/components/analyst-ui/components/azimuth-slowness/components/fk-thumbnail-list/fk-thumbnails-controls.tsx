/* eslint-disable react/destructuring-assignment */
import { Menu, MenuDivider } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { MenuItem2 } from '@blueprintjs/popover2';
import { DeprecatedToolbar, DeprecatedToolbarTypes } from '@gms/ui-core-components';
import React from 'react';

const MARGINS_FOR_TOOLBAR_PX = 0;
/**
 * Fk Thumbnails Controls Props
 */
export interface FkThumbnailsControlsProps {
  currentFilter: FilterType;
  widthPx: number;
  anyDisplayedFksNeedReview: boolean;
  onlyOneFkIsSelected: boolean;
  anyUnassociatedSelected: boolean;
  updateFkThumbnail(px: number): void;
  updateFkFilter(filter: FilterType): void;
  clearSelectedUnassociatedFks();
  nextFk(): void;
}

/**
 * Different filters that are available
 */
export enum FilterType {
  firstP = 'First P',
  all = 'All',
  needsReview = 'Needs review'
}

/**
 * Pixels widths of available thumbnail sizes
 */
export enum FkThumbnailSize {
  SMALL = 70,
  MEDIUM = 110,
  LARGE = 150
}

/**
 * FK Thumbnails Controls Component
 * Filtering / review controls for the FK
 */
export function FkThumbnailsControls(props: FkThumbnailsControlsProps) {
  const toolbarItemsLeft: DeprecatedToolbarTypes.ToolbarItem[] = [];
  toolbarItemsLeft.push({
    rank: 1,
    label: 'Filter',
    type: DeprecatedToolbarTypes.ToolbarItemType.Dropdown,
    tooltip: 'Filter the fks',
    dropdownOptions: FilterType,
    widthPx: 130,
    value: props.currentFilter,
    onChange: value => props.updateFkFilter(value as FilterType)
  });
  const fkThumbnailMenuPopup = (
    <Menu>
      <MenuDivider title="Thumbnail Size" className="menu-title" />
      <MenuItem2
        icon={IconNames.SMALL_SQUARE}
        onClick={() => props.updateFkThumbnail(FkThumbnailSize.SMALL)}
        text="Small"
      />
      <MenuItem2
        icon={
          // TODO: This should be created into a separate svg file and used
          <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16">
            <path
              id="med-square"
              d="M2.067,10.06H10.06V2.067H2.067ZM1.213,0A1.146,1.146,0,0,0,0,1.213v9.7a1.146,1.146,0,0,0,1.213,1.213h9.7a1.146,1.146,0,0,0,1.212-1.213v-9.7A1.146,1.146,0,0,0,10.915,0Z"
              transform="translate(1.75 1.75)"
              fill="#bfccd6"
              fillRule="evenodd"
            />
          </svg>
        }
        onClick={() => props.updateFkThumbnail(FkThumbnailSize.MEDIUM)}
        text="Medium"
      />
      <MenuItem2
        icon={IconNames.SQUARE}
        onClick={() => props.updateFkThumbnail(FkThumbnailSize.LARGE)}
        text="Large"
      />
      <MenuDivider title="Approval" className="menu-title" />
      <MenuItem2
        icon={IconNames.CONFIRM}
        disabled={false}
        onClick={() => props.clearSelectedUnassociatedFks()}
        text="Accept selected"
      />
      <MenuItem2
        icon={IconNames.CROSS}
        disabled={false}
        onClick={() => props.clearSelectedUnassociatedFks()}
        text="Hide selected"
      />
    </Menu>
  );
  const toolbarItems: DeprecatedToolbarTypes.ToolbarItem[] = [];
  const nextButton: DeprecatedToolbarTypes.ButtonItem = {
    rank: 1,
    label: 'Next',
    tooltip: 'Selected next fk that needs review',
    type: DeprecatedToolbarTypes.ToolbarItemType.Button,
    // eslint-disable-next-line @typescript-eslint/unbound-method
    onClick: props.nextFk,
    disabled: !props.anyDisplayedFksNeedReview || !props.onlyOneFkIsSelected,
    widthPx: 50
  };
  const optionsPopover: DeprecatedToolbarTypes.PopoverItem = {
    rank: 2,
    label: 'Options',
    type: DeprecatedToolbarTypes.ToolbarItemType.Popover,
    tooltip: 'Options for displaying fk thumbnails',
    popoverContent: fkThumbnailMenuPopup,
    widthPx: 30,
    onlyShowIcon: true,
    icon: IconNames.MORE,
    onPopoverDismissed: () => {
      // This empty arrow function is intentional.  This comment satisfies removing a SonarQube's critical issue
    }
  };
  toolbarItems.push(nextButton);
  toolbarItems.push(optionsPopover);

  return (
    <div className="azimuth-slowness-thumbnails-controls__wrapper">
      <DeprecatedToolbar
        itemsLeft={toolbarItemsLeft}
        toolbarWidthPx={props.widthPx - MARGINS_FOR_TOOLBAR_PX}
        itemsRight={toolbarItems}
      />
    </div>
  );
}
