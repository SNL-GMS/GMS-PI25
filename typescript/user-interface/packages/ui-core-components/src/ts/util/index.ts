export * from './display-util';
export * from './non-ideal-state';
export * from './validation';
