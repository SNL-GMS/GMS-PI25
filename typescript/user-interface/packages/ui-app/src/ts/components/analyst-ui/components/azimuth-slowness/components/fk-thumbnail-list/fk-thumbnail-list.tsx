/* eslint-disable react/destructuring-assignment */
import type { EventTypes, FkTypes, SignalDetectionTypes } from '@gms/common-model';
import type { FkChannelSegmentRecord } from '@gms/ui-state';
import { useGetFkChannelSegments } from '@gms/ui-state';
import type Immutable from 'immutable';
import findIndex from 'lodash/findIndex';
import findLastIndex from 'lodash/findLastIndex';
import React, { useEffect } from 'react';

import { getFkData, getFkUnitForSdId } from '../../../../common/utils/fk-utils';
import { fkNeedsReview } from '../fk-util';
import { FkThumbnailContainer } from './fk-thumbnail-container';

/**
 * Fk Thumbnails Props
 */
export interface FkThumbnailListProps {
  sortedSignalDetections: SignalDetectionTypes.SignalDetection[];
  unassociatedSignalDetections: SignalDetectionTypes.SignalDetection[];
  signalDetectionIdsToFeaturePrediction: Immutable.Map<string, EventTypes.FeaturePrediction[]>;
  thumbnailSizePx: number;
  fkUnitsForEachSdId: Immutable.Map<string, FkTypes.FkUnits>;
  arrivalTimeMovieSpectrumIndex: number;
  displayedSignalDetection: SignalDetectionTypes.SignalDetection;
  selectedSdIds: string[];
  setDisplayedSignalDetection(sd: SignalDetectionTypes.SignalDetection): void;
  setSelectedSdIds(sdIds: string[]): void;
  showFkThumbnailContextMenu(event: React.MouseEvent): void;
}

/**
 * List of fk thumbnails with controls for filtering them
 */
export function FkThumbnailList(props: FkThumbnailListProps) {
  const fkChannelSegments: FkChannelSegmentRecord = useGetFkChannelSegments();
  /**
   * Automatically selects an fk when an event is loaded
   */
  useEffect(() => {
    // If no displayedSignalDetection is selected or
    // displayedSignalDetection is no longer in sortedSignalDetections list
    if (
      !props.displayedSignalDetection ||
      !props.sortedSignalDetections.find(sd => sd.id === props.displayedSignalDetection.id)
    ) {
      // Walk thru the list to find first that needs review
      let selectedSd = props.sortedSignalDetections
        .filter(
          sd =>
            !props.unassociatedSignalDetections.find(unassociatedSd => unassociatedSd.id === sd.id)
        )
        .find(sd => fkNeedsReview(sd, getFkData(sd, fkChannelSegments)));

      // If not found from needs review then take first available with FkData
      if (!selectedSd) {
        selectedSd = props.sortedSignalDetections.find(
          sd => getFkData(sd, fkChannelSegments) !== undefined
        );
      }

      // If found a new one set the selected id
      if (selectedSd) {
        props.setSelectedSdIds([selectedSd.id]);
        props.setDisplayedSignalDetection(selectedSd);
      }
    } else if (props.displayedSignalDetection && props.selectedSdIds.length === 0) {
      props.setSelectedSdIds([props.displayedSignalDetection.id]);
    } else if (props.sortedSignalDetections.length === 0) {
      props.setSelectedSdIds([]);
      props.setDisplayedSignalDetection(undefined);
    }
  }, [fkChannelSegments, props]);

  /**
   * onKeyDown event handler for selecting multiple fk at once
   */
  function onKeyDown(e: React.KeyboardEvent<HTMLDivElement>): void {
    if (e.repeat) {
      return;
    }

    if (e.shiftKey && (e.ctrlKey || e.metaKey) && e.key.toLowerCase() === 'a') {
      // Shift + CmrOrCtrl + a ==> add all to current selection
      const selectedIds: string[] = props.selectedSdIds;
      props.sortedSignalDetections.forEach(sd => {
        if (props.selectedSdIds.indexOf(sd.id) === -1) {
          selectedIds.push(sd.id);
        }
      });
      props.setSelectedSdIds(selectedIds);
    } else if ((e.ctrlKey || e.metaKey) && e.key.toLowerCase() === 'a') {
      // CmrOrCtrl + a ==> select all
      const selectedIds: string[] = [...props.sortedSignalDetections.map(sd => sd.id)];
      props.setSelectedSdIds(selectedIds);
    } else if (e.key.toLowerCase() === 'escape') {
      // Escape ==> deselect all
      props.setSelectedSdIds([]);
      props.setDisplayedSignalDetection(undefined);
    }
  }

  /**
   * Builds the selected SD id list
   *
   * @param sdId selected SD
   * @returns selected SD id list
   */
  function buildMultiSelectedIds(sdId: string): string[] {
    // shift range selection
    const selectedIds = [...props.selectedSdIds];
    const fkIds: string[] = props.sortedSignalDetections.map(sd => sd.id);
    const selectedIndex: number = fkIds.indexOf(sdId);
    const minIndex: number = findIndex(fkIds, i => selectedIds.indexOf(i) !== -1);
    const maxIndex: number = findLastIndex(fkIds, i => selectedIds.indexOf(i) !== -1);
    fkIds.forEach(i => {
      if (selectedIds.indexOf(i) === -1) {
        const index: number = fkIds.indexOf(i);
        if (
          (index >= selectedIndex && index < minIndex) ||
          (index > maxIndex && index <= selectedIndex)
        ) {
          selectedIds.push(i);
        }
      }
    });
    return selectedIds;
  }

  /**
   * Selects a slicked thumbnail
   */
  function onThumbnailClick(e: React.MouseEvent<HTMLCanvasElement>, id: string): void {
    // Ignore signal detections where the FK is not available
    const sdToAdd = props.sortedSignalDetections.find(s => s.id === id);
    if (!getFkData(sdToAdd, fkChannelSegments)) {
      return;
    }
    let selectedIds: string[];
    if (e.shiftKey && props.selectedSdIds.length > 0) {
      selectedIds = buildMultiSelectedIds(id);
    } else if (e.ctrlKey || e.metaKey) {
      // add/remove to current selection
      selectedIds = [...props.selectedSdIds];

      if (selectedIds.indexOf(id) >= 0) {
        selectedIds = selectedIds.splice(selectedIds.indexOf(id), 1);
      } else {
        selectedIds.push(id);
      }
    } else {
      selectedIds = [id];
      props.setDisplayedSignalDetection(props.sortedSignalDetections.find(sd => sd.id === id));
    }
    props.setSelectedSdIds(selectedIds);
  }

  /**
   * Renders the component.
   */
  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      className="azimuth-slowness-thumbnails__wrapper-3"
      onKeyDown={onKeyDown}
      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
      tabIndex={0}
      key="fk-thumbnails"
    >
      {props.sortedSignalDetections.map(sd => (
        <FkThumbnailContainer
          key={sd.id}
          data={sd}
          signalDetectionFeaturePredictions={
            props.signalDetectionIdsToFeaturePrediction.has(sd.id)
              ? props.signalDetectionIdsToFeaturePrediction.get(sd.id)
              : []
          }
          sizePx={props.thumbnailSizePx}
          selected={props.selectedSdIds.indexOf(sd.id) >= 0}
          onClick={(e: React.MouseEvent<HTMLCanvasElement>) => onThumbnailClick(e, sd.id)}
          fkUnit={getFkUnitForSdId(sd.id, props.fkUnitsForEachSdId)}
          isUnassociated={
            props.unassociatedSignalDetections.find(unSd => unSd.id === sd.id) !== undefined
          }
          // eslint-disable-next-line @typescript-eslint/unbound-method
          showFkThumbnailMenu={props.showFkThumbnailContextMenu}
          arrivalTimeMovieSpectrumIndex={props.arrivalTimeMovieSpectrumIndex}
          selectedFk={getFkData(sd, fkChannelSegments)}
        />
      ))}
    </div>
  );
}
