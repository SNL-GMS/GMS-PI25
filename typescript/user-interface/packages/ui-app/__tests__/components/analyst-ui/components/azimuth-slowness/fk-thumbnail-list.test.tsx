/* eslint-disable react/jsx-props-no-spreading */
import type { EventTypes, FkTypes } from '@gms/common-model';
import { signalDetectionsData } from '@gms/common-model/__tests__/__data__';
import { getStore } from '@gms/ui-state';
import { getTestFkChannelSegmentRecord } from '@gms/ui-state/__tests__/__data__';
import { render } from '@testing-library/react';
import { act } from '@testing-library/react-hooks';
import Immutable from 'immutable';
import * as React from 'react';
import { Provider } from 'react-redux';
import * as util from 'util';

import type { FkThumbnailListProps } from '../../../../../src/ts/components/analyst-ui/components/azimuth-slowness/components/fk-thumbnail-list/fk-thumbnail-list';
import { FkThumbnailList } from '../../../../../src/ts/components/analyst-ui/components/azimuth-slowness/components/fk-thumbnail-list/fk-thumbnail-list';

Object.defineProperty(window, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(window, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});
Object.defineProperty(global, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(global, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});

// This is required so that jest.spyOn doesn't throw a type error
jest.mock('@gms/ui-state', () => {
  const actual = jest.requireActual('@gms/ui-state');
  return {
    ...actual,
    useGetFkChannelSegments: jest.fn().mockReturnValue(() => {
      return getTestFkChannelSegmentRecord(signalDetectionsData[0]);
    })
  };
});
// TODO the file name .redo. makes the tests skip everything in here, the tests need to be redone
describe('FK thumbnails tests', () => {
  const mockProps: FkThumbnailListProps = {
    sortedSignalDetections: signalDetectionsData,
    signalDetectionIdsToFeaturePrediction: Immutable.Map<string, EventTypes.FeaturePrediction[]>(),
    thumbnailSizePx: 300,
    selectedSdIds: [signalDetectionsData[0].id],
    unassociatedSignalDetections: [signalDetectionsData[0]],
    arrivalTimeMovieSpectrumIndex: 0,
    fkUnitsForEachSdId: Immutable.Map<string, FkTypes.FkUnits>(),
    displayedSignalDetection: signalDetectionsData[0],
    showFkThumbnailContextMenu: () => {
      /** empty */
    },
    setSelectedSdIds: () => {
      /** empty */
    },
    setDisplayedSignalDetection: () => {
      /** empty */
    }
  };

  const store = getStore();
  it('matches snapshot', () => {
    let container;
    act(() => {
      container = render(
        <Provider store={store}>
          <FkThumbnailList {...mockProps} />
        </Provider>
      );
    });
    expect(container).toMatchSnapshot();
  });
});
