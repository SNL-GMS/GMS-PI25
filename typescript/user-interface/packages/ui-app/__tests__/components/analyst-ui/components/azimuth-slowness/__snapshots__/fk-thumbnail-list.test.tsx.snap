// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FK thumbnails tests matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="azimuth-slowness-thumbnails__wrapper-3"
        tabindex="0"
      >
        <div
          class="fk-thumbnail selected"
          style="width: 300px; height: 300px;"
        >
          <div
            class="fk-thumbnail__label--reviewed"
          >
            ASAR P
          </div>
          <canvas
            class="fk-thumbnail__canvas"
            height="300"
            width="300"
          />
        </div>
        <div
          class="fk-thumbnail "
          style="width: 300px; height: 300px;"
        >
          <div
            class="fk-thumbnail__label--reviewed"
          >
            ASAR Pg
          </div>
          <canvas
            class="fk-thumbnail__canvas"
            height="300"
            width="300"
          />
        </div>
        <div
          class="fk-thumbnail "
          style="width: 300px; height: 300px;"
        >
          <div
            class="fk-thumbnail__label--reviewed"
          >
            ASAR P
          </div>
          <canvas
            class="fk-thumbnail__canvas"
            height="300"
            width="300"
          />
        </div>
        <div
          class="fk-thumbnail "
          style="width: 300px; height: 300px;"
        >
          <div
            class="fk-thumbnail__label--reviewed"
          >
            PDAR P
          </div>
          <canvas
            class="fk-thumbnail__canvas"
            height="300"
            width="300"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="azimuth-slowness-thumbnails__wrapper-3"
      tabindex="0"
    >
      <div
        class="fk-thumbnail selected"
        style="width: 300px; height: 300px;"
      >
        <div
          class="fk-thumbnail__label--reviewed"
        >
          ASAR P
        </div>
        <canvas
          class="fk-thumbnail__canvas"
          height="300"
          width="300"
        />
      </div>
      <div
        class="fk-thumbnail "
        style="width: 300px; height: 300px;"
      >
        <div
          class="fk-thumbnail__label--reviewed"
        >
          ASAR Pg
        </div>
        <canvas
          class="fk-thumbnail__canvas"
          height="300"
          width="300"
        />
      </div>
      <div
        class="fk-thumbnail "
        style="width: 300px; height: 300px;"
      >
        <div
          class="fk-thumbnail__label--reviewed"
        >
          ASAR P
        </div>
        <canvas
          class="fk-thumbnail__canvas"
          height="300"
          width="300"
        />
      </div>
      <div
        class="fk-thumbnail "
        style="width: 300px; height: 300px;"
      >
        <div
          class="fk-thumbnail__label--reviewed"
        >
          PDAR P
        </div>
        <canvas
          class="fk-thumbnail__canvas"
          height="300"
          width="300"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
